### PR TITLE
Shoulder Gun Module

### PR DESCRIPTION
**Pull request by hyper2snyper**
## About The Pull Request
Adds a new gun attachment/armor module that is a t3-4 drop from the researcher console.
This attachment, goes onto a gun. Once on a guns rail slot you can attach the gun to your modular armor as if it was an armor attachment. Once its attached you can reload the mounted gun by right clicking it with the ammo in hand. Right clicking with an empty hand will allow you to select whether the gun fires on right/left/middle click. The unique action keybind will activate a shoulder gun as well.

The shoulder gun can not be fired while the user is holding another gun or a shield.

~~It is now available from req for 30 points.~~

## Why It's Good For The Game
Really cool researcher content.
Using t4 research material I only acquired one after using about 8 materials. I doubt this will have much effect on balance.
It is a module so no other armor modules are compatible.

## Changelog
:cl:
add: Adds the Shoulder Gun Module, a new gun attachment/armor module that is a t3-4 drop from the researcher console.
/:cl:

### CODE REVIEW

**Comment by jroinc** [comment]
Upping unga DPS with additional wield slots for guns is a bad precedent. They really, really don't need shouldershotguns or more DPS with no wield requirement.

I am aware that it's hard to acquire in this PR, and I'm not opposed to that, it'll just open the gates for the inevitable "add it to req" PR.

**Comment by Lumipharon** [comment]
Triple wielding sounds hilarious.
And you do have two shoulders...

**Comment by Walarks** [comment]
I think a lightweight shoulder turret would be interesting

**Comment by DarkWolfKai** [comment]
Add a arm blade for it and boom, Were the Marine Slayer, Pretty good concept Bird.

**Comment by AndroidMK2** [comment]
Add predator shoulder turrets 😳

**Review by BraveMole — CHANGES_REQUESTED** [review]
Still a hard no for me.

Maybe for a flare launcher, or a pepper gun.

Revolvers, flamers, pistols, rifles, laserrifle, shotguns (SHOTGUNS???!), smgs,  snipers, no.

**Comment by ThelastArchiver** [comment]
imagine shoulder mounted rocket launcher

**Comment by Moggram** [comment]
> 
> 
> imagine shoulder mounted rocket launcher

Ya, no. I am going to side with BraveMole here. This is right in the ball-park of over powered.

**Comment by Tennessee116** [comment]
Possible modification: shoulder-mounted spears

File: code/modules/projectiles/gun_attachables.dm [hunk 1]
```diff
@@ -1490,19 +1490,160 @@ inaccurate. Don't worry if force is ever negative, it won't runtime.
...
+		UnregisterSignal(user, COMSIG_MOB_MOUSEDOWN)
+		master_gun.set_gun_user(null)
+	else
+		ENABLE_BITFIELD(master_gun.flags_item, IS_DEPLOYED)
+		overlays += image('icons/Marine/marine-weapons.dmi', src, "active")
+		update_icon()
+		master_gun.set_gun_user(user)
+		RegisterSignal(user, COMSIG_MOB_MOUSEDOWN, .proc/handle_firing)
+		master_gun.RegisterSignal(user, COMSIG_MOB_MOUSEDRAG, /obj/item/weapon/gun.proc/change_target)
+	for(var/action_to_update in actions)
```
Comment (BraveMole) [inline code/modules/projectiles/gun_attachables.dm:1555]: AS in

File: code/modules/projectiles/gun_system.dm [hunk 2]
```diff
@@ -826,7 +826,12 @@ and you're good to go.
 
 
 	play_fire_sound(loc)
-	muzzle_flash(firing_angle, master_gun ? gun_user : loc)
+	var/flash_loc
+	if(master_gun || !istype(loc, /obj/machinery/deployable/mounted))
+		flash_loc = gun_user
+	else
+		flash_loc = loc
+	muzzle_flash(firing_angle, flash_loc)
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:834]: use ?

File: code/__DEFINES/attachments.dm [hunk 3]
```diff
@@ -25,6 +25,7 @@
 #define ATTACH_APPLY_ON_MOB (1<<3)
 #define ATTACH_SEPERATE_MOB_OVERLAY (1<<4)
 #define ATTACH_NO_HANDS (1<<5)
+#define ATTACH_BYPASS_ALLOWED_LIST (1<<6)
```
Comment (BraveMole) [inline code/__DEFINES/attachments.dm:28]: what is this doing?

**Review by BraveMole — CHANGES_REQUESTED** [review]
you should not be able to PB with this gun IMO

**Review by Kuro020 — APPROVED** [review]
sprites aokay even though this pr is gonna get mole'd kekw

**Comment by Tristan63** [comment]
stinky

File: code/modules/projectiles/gun_attachables.dm [hunk 4]
```diff
@@ -1490,19 +1490,159 @@ inaccurate. Don't worry if force is ever negative, it won't runtime.
...
+	battery = detaching_gun.sentry_battery
 	battery?.forceMove(src)
-	detaching_item.sentry_battery = null
-	detaching_item.RemoveElement(/datum/element/deployable_item, /obj/machinery/deployable/mounted/sentry/buildasentry, deploy_time, undeploy_time)
+	detaching_gun.sentry_battery = null
+	detaching_gun.RemoveElement(/datum/element/deployable_item, /obj/machinery/deployable/mounted/sentry/buildasentry, deploy_time, undeploy_time)
 
 
+/obj/item/attachable/shoulder_mount
+	name = "Experimental Shoulder Attachment Point"
```
Comment (XSlayer300) [inline code/modules/projectiles/gun_attachables.dm:1507]:
```suggestion
	name = "experimental shoulder attachment point"
```

File: code/modules/projectiles/gun_attachables.dm [hunk 5]
```diff
@@ -1490,19 +1490,159 @@ inaccurate. Don't worry if force is ever negative, it won't runtime.
...
+	desc = "A brand new advance in combat technology. This device, once attached to a firearm, will allow the firearm to be mounted onto any piece of modular armor. Once attached to the armor and activated, the gun will fire when the user chooses. Once attached to the armor, right clicking the armor with an empty hand will select what click will fire the armor (middle, right, left). Right clicking with ammunition will reload the gun. Using the unique action keybind will perform the guns unique action only when the gun is active."
```
Comment (XSlayer300) [inline code/modules/projectiles/gun_attachables.dm:1508]:
```suggestion
	desc = "A brand new advance in combat technology. This device, once attached to a firearm, will allow the firearm to be mounted onto any piece of modular armor. Once attached to the armor and activated, the gun will fire when the user chooses.\nOnce attached to the armor, <b>right clicking</b> the armor with an empty hand will select what click will fire the armor (middle, right, left). <b>Right clicking</b> with ammunition will reload the gun. Using the <b>Unique Action</b> keybind will perform the weapon's unique action only when the gun is active."
```

File: code/datums/elements/attachment.dm [hunk 6]
```diff
@@ -34,6 +34,10 @@
 
 	RegisterSignal(target, COMSIG_ITEM_IS_ATTACHING, .proc/on_attaching)
 
+/datum/element/attachment/Detach(datum/source, force)
+	UnregisterSignal(source, COMSIG_ITEM_IS_ATTACHING)
+	. = ..()
```
Comment (TiviPlus) [inline code/datums/elements/attachment.dm:39]:
```suggestion
/datum/element/attachment/Detach(datum/source, force)
	UnregisterSignal(source, COMSIG_ITEM_IS_ATTACHING)
	return ..()
```

File: code/modules/projectiles/gun_attachables.dm [hunk 7]
```diff
@@ -1490,19 +1490,159 @@ inaccurate. Don't worry if force is ever negative, it won't runtime.
...
+///Reloads the gun
+/obj/item/attachable/shoulder_mount/proc/reload_gun(datum/source, obj/item/attacking_item, mob/living/user)
+	SIGNAL_HANDLER
+	INVOKE_ASYNC(master_gun, /obj/item/weapon/gun.proc/reload, user, attacking_item)
+
+///Performs the unique action after firing.
+/obj/item/attachable/shoulder_mount/proc/after_fire(datum/source, atom/target, obj/item/weapon/gun/fired_gun)
+	SIGNAL_HANDLER
+	master_gun.cock(master_gun.gun_user)
 /obj/item/attachable/flamer_nozzle
```
Comment (TiviPlus) [inline code/modules/projectiles/gun_attachables.dm:1646]:
```suggestion
	SIGNAL_HANDLER
	master_gun.cock(master_gun.gun_user)

/obj/item/attachable/flamer_nozzle
```

**Review by TiviPlus — CHANGES_REQUESTED** [review]
and slayers grammer changes pls

**Comment by vilereaver41** [comment]
i'm willing to give anything atleast a chance despite the opness this might be i mean why not. you get some laughs from it.

**Comment by Tristan63** [comment]
could make it a req only item for like 40 points

**Comment by hyper2snyper** [comment]
> could make it a req only item for like 40 points

Well. That was the original plan. However this module mainly has coolness going for it. therefore it is best suited for research content

**Review by BraveMole — APPROVED** [review]

**Comment by Lumipharon** [comment]
It's quite dirty with sadar etc, since it solves storage issues, and effectively means you have 2 magharnessed guns. If it's locked behind research though it will probably be fine.

**Review by BraveMole — APPROVED** [review]

**Pull request merged**